src/playground/services/invoices/delete/endpoint.clj

```diff
@@ -3,22 +3,29 @@
    [clojure.spec.alpha :as spec]
    [clojure.java.jdbc :as jdbc]
    [honeysql.core :as h]
+   [buddy.auth :refer [authenticated? throw-unauthorized]]
+   [io.pedestal.http.route :refer (url-for)]
    [playground.services.invoices.delete.logic :as logic]
    [playground.services.invoices.retrieve-all.logic :as retrieve-all.logic]
-   [playground.models.user :as models.user]))
+   [playground.models.user :as models.user]
+   [ring.util.response :as ring-resp]))
 
 (spec/def ::api (spec/keys :req-un [::models.user/id]))
 
-(defn perform [{{:keys [id]} :path-params :keys [db] :as request}]
-  (let [db (->> db :pool (hash-map :datasource))
-        _  (->> (logic/to-delete id)
-                (h/format)
-                (jdbc/execute! db))
-        fetch  (h/format (retrieve-all.logic/query-all))
-        left   (->> (logic/to-serialize)
+(defn perform [{{:keys [id]} :path-params :keys [db session] :as request}]
+  (let  [role (get-in session [:identity :role])]
+    (if (= "admin" role)
+      (let [db (->> db :pool (hash-map :datasource))
+            _  (->> (logic/to-delete id)
                     (h/format)
-                    (jdbc/query db)
-                    (first))
-        result (jdbc/query db fetch)
-        result-map {:result result :deleted id}]
-    {:status 302 :headers {"Location" "/invoices"} :body "" :flash (str "Entry " id " has beed deleted.")}))
+                    (jdbc/execute! db))
+            fetch  (h/format (retrieve-all.logic/query-all))
+            left   (->> (logic/to-serialize)
+                        (h/format)
+                        (jdbc/query db)
+                        (first))
+            result (jdbc/query db fetch)
+            result-map {:result result :deleted id}]
+        (-> (ring-resp/redirect (url-for :invoices))
+            (assoc :flash (str "Entry " id " has beed deleted."))))
+      (throw-unauthorized))))
```

Comment:
In `if`s, when a 'branch' (else/then) is much bigger than the other, the smaller one should go first. That way one keeps less context in his head when reading code

Consequently, use `if-not` and swap the branches

Reply:
(this suggestion will be unnecessary given the authorization interceptor, but please remember the `if` rule for future occasions)